File: sources/fortran/04_array_sum/main.f90
! task 04 array_sum — expected output: 499999500000
! build: gfortran -O3 -o prog main.f90 (flang -O3 -o prog main.f90)    run: ./prog

program main
  implicit none
  integer, parameter :: n = 1000000
  integer(kind=8), allocatable :: a(:)
  integer(kind=8) :: total, i

  allocate(a(0:n-1))
  do i = 0_8, int(n, kind=8) - 1_8
    a(i) = i
  end do

  total = 0_8
  do i = 0_8, int(n, kind=8) - 1_8
    total = total + a(i)
  end do

  write(*,'(i0)') total
end program main

File: sources/pascal/04_array_sum/main.pas
{ task 04 array_sum — expected output: 499999500000 }
{ build: fpc -O3 -oprogram main.pas    run: ./program }
{ Windows x64: i386-win32 fpc + cross.x86_64-win64 add-on, build with -Px86_64; run as program.exe }

program prog;
{$mode objfpc}{$H+}

uses
  SysUtils;

var
  arr: array of Int64;
  i: Integer;
  total: Int64;
begin
  SetLength(arr, 1000000);
  for i := 0 to 999999 do
    arr[i] := i;
  total := 0;
  for i := 0 to 999999 do
    total := total + arr[i];
  WriteLn(total);
end.
